k2/csrc/intersect_dense.cu

```diff
@@ -0,0 +1,848 @@
+            // comes from appending a_fsas_.shape and incoming_arcs_.shape along
+            // axis 1. So (2 * this_fsa_first_arc) is the start of this FSA's
+            // data, and adding this_fsa_num_arcs means skipping over the arcs
+            // arranged as in a_fsas_.shape.
+            carc.incoming_arc_idx012 =
+                2 * this_fsa_first_arc + this_fsa_num_arcs + incoming_arc_idx12;
+          }
+          carcs_data[i] = carc;
+        });
+  }
+
+  void InitFsaInfo() {
+    NVTX_RANGE(K2_FUNC);
+    int32_t *b_fsas_row_splits1_data = b_fsas_.shape.RowSplits(1).Data(),
+            *a_fsas_row_splits1_data = a_fsas_.shape.RowSplits(1).Data(),
+            *a_fsas_row_splits2_data = a_fsas_.shape.RowSplits(2).Data();
+    int32_t scores_stride = b_fsas_.scores.ElemStride0();
+
+    fsa_info_ = Array1<FsaInfo>(c_, num_fsas_ + 1);
+    FsaInfo *fsa_info_data = fsa_info_.Data();
+    int32_t num_fsas = num_fsas_;
+    K2_EVAL(
+        c_, num_fsas_ + 1, lambda_set_fsa_info, (int32_t i)->void {
+          FsaInfo info;
+          if (i < num_fsas) {
+            info.T = uint16_t(b_fsas_row_splits1_data[i + 1] -
+                              b_fsas_row_splits1_data[i]);
+            info.num_states = uint16_t(a_fsas_row_splits1_data[i + 1] -
+                                       a_fsas_row_splits1_data[i]);
+          } else {
+            info.T = 0;
+            info.num_states = 0;
+          }
+          info.scores_offset = b_fsas_row_splits1_data[i] * scores_stride;
+          info.state_offset = a_fsas_row_splits1_data[i];
+          info.arc_offset = a_fsas_row_splits2_data[info.state_offset];
+          fsa_info_data[i] = info;
+        });
+  }
+
+  /*
+    InitSteps() sets up steps_; it works out metadata and allocates memory, but
+    does not do any of the actual computation.
+   */
+  void InitSteps() {
+    NVTX_RANGE(K2_FUNC);
+    // This vector, of length num_fsas_, tells us how many copies of (the states
+    // of the i'th decoding graph) we have.  It equals (the length of the
+    // sequence of log-likes in b_fsas_) + 1.  It is monotonically decreasing
+    // (thanks to how we require the FSAs to be sorted).
+    Array1<int32_t> num_copies_per_fsa(c_, num_fsas_);
+    const int32_t *b_row_splits_data = b_fsas_.shape.RowSplits(1).Data();
+    int32_t *num_copies_per_fsa_data = num_copies_per_fsa.Data();
+
+    K2_EVAL(
+        c_, num_fsas_, lambda_set_num_copies, (int32_t i)->void {
+          num_copies_per_fsa_data[i] =
+              1 + b_row_splits_data[i + 1] - b_row_splits_data[i];
+        });
+
+    std::vector<int32_t> range(num_fsas_);
+    // fill with 1, 2, .. num_fsas_.
+    std::iota(range.begin(), range.end(), 1);
+    std::vector<RaggedShape> arc_scores_prefixes =
+        GetPrefixes(arc_scores_.shape, range);
+
+    ContextPtr c_cpu = GetCpuContext();
+
+    // This vector, of length T_ + 1, tells us, for each frame 0 <= t <= T, how
```

Comment:
> This vector, of length T_ + 1

It is not necessary of length `T_ + 1`.

`T_` is the longest sequence in the DenseFsaVec, **BUT** it may happen
that this sequence is never used. That is, `a_to_b_map` excludes this
sequence.

When this case happens,   line 562 will cause out-of-bound error,
like the one in https://github.com/k2-fsa/k2/issues/693